File: plugin/zeppelin/src/main/scala/jetbrains/zeppelin/idea/settings/interpreter/InterpreterSettingsDialog.scala
package jetbrains.zeppelin.idea.settings.interpreter

import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogWrapper
import javax.swing.{JComponent, SwingConstants}
import jetbrains.zeppelin.components.ZeppelinComponent
import jetbrains.zeppelin.models.{Dependency, InstantiationType, Interpreter, InterpreterOption}

import scala.collection.JavaConverters._

class InterpreterSettingsDialog(project: Project, var interpreter: Interpreter) extends DialogWrapper(project) {
  val title = s"${interpreter.name} interpreter settings"
  private val myPanel = new InterpreterSettingsForm()

  setTitle(title)
  setButtonsAlignment(SwingConstants.CENTER)
  init()

  override def createCenterPanel(): JComponent = myPanel.getContentPane

  override def doOKAction(): Unit = {
    val newDependencies: List[Dependency] = getNewDependencies
    val newOptions = getNewOptions

    interpreter = interpreter.copy(dependencies = newDependencies, option = newOptions)

    val connection = ZeppelinComponent.connectionFor(project)
    val actionService = connection.service
    actionService.updateInterpreterSettings(interpreter)

    UpdateInterpreterHandler.getAll.foreach(_.updateInterpreter(project))
    super.doOKAction()
  }

  override def init(): Unit = {
    super.init()
    updateDependencyList()
    updateInstantiationType()
  }

  def updateDependencyList(): Unit = {
    val dependencies = interpreter.dependencies.map(_.groupArtifactVersion)
    myPanel.initDataModel(dependencies.asJava)
  }

  def updateInstantiationType(): Unit = {
    val options = interpreter.option
    val values: List[String] = InstantiationType.values.toList.map(_.toString)
    myPanel.initInstantiationTypes(values.asJava, options)
  }

  private def getNewDependencies = {
    val dependenciesNames = myPanel.getModelList.asScala.toList
    val originalDependencies = interpreter.dependencies
    val newDependencies = dependenciesNames
      .filter(_ != null)
      .map(it => {
        originalDependencies
          .find(_.groupArtifactVersion == it)
          .getOrElse(Dependency(it))
      })
    newDependencies
  }

  private def getNewOptions = {
    if (myPanel.isGlobally) {
      InterpreterOption()
    } else {
      InterpreterOption(Some(myPanel.getPerNoteValue), Some(myPanel.getPerUserValue))
    }
  }
}

File: zeppelin/src/org/intellij/plugin/zeppelin/idea/settings/interpreter/InterpreterSettingsDialog.kt
package org.intellij.plugin.zeppelin.idea.settings.interpreter

import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogWrapper
import org.intellij.plugin.zeppelin.components.ZeppelinComponent
import org.intellij.plugin.zeppelin.extensionpoints.UpdateInterpreterHandler
import org.intellij.plugin.zeppelin.models.Dependency
import org.intellij.plugin.zeppelin.models.InstantiationType
import org.intellij.plugin.zeppelin.models.Interpreter
import org.intellij.plugin.zeppelin.models.InterpreterOption
import org.intellij.plugin.zeppelin.service.ZeppelinActionService
import javax.swing.JComponent

open class InterpreterSettingsDialog(private val project: Project,
                                     var interpreter: Interpreter) : DialogWrapper(project) {

    override fun getTitle(): String = "${interpreter.name} interpreter settings"

    private val myPanel: InterpreterSettingsForm = InterpreterSettingsForm()
    override fun createCenterPanel(): JComponent = myPanel.contentPane
    override fun doOKAction() {
        val newDependencies: List<Dependency> = getNewDependencies()
        val newOptions: InterpreterOption = getNewOptions()
        interpreter = interpreter.copy(dependencies = newDependencies, option = newOptions)
        val connection: ZeppelinComponent = ZeppelinComponent.connectionFor(project)
        val actionService: ZeppelinActionService = connection.service
        actionService.updateInterpreterSettings(interpreter)
        UpdateInterpreterHandler.getAll().forEach { it.updateInterpreter(project) }
        super.doOKAction()
    }

    override fun init() {
        super.init()
        updateDependencyList()
        updateInstantiationType()
    }

    private fun updateDependencyList() {
        val dependencies = interpreter.dependencies.map {it.groupArtifactVersion}
        myPanel.initDataModel(dependencies)
    }

    private fun updateInstantiationType() {
        val options: InterpreterOption = interpreter.option
        val values: List<String> = InstantiationType.values().toList().map { it.toString() }
        myPanel.initInstantiationTypes(values, options)
    }

    private fun getNewDependencies(): List<Dependency> {
        val dependenciesNames: List<String?> = myPanel.modelList.toList()
        val originalDependencies: List<Dependency> = interpreter.dependencies
        return dependenciesNames.filter { it != null }
            .map { it -> originalDependencies.find { origIt -> origIt.groupArtifactVersion == it } ?: Dependency(it!!) }
    }

    private fun getNewOptions(): InterpreterOption {
        return if (myPanel.isGlobally) {
            InterpreterOption()
        } else {
            InterpreterOption(InstantiationType.valueOf(myPanel.perNoteValue),
                              InstantiationType.valueOf(myPanel.perUserValue))
        }
    }
}